src/storage/src/flush/scheduler.rs

```diff
@@ -0,0 +1,234 @@
+    #[inline]
+    fn key(&self) -> FlushKey {
+        (self.shared.id(), self.flush_sequence)
+    }
+
+    fn complete(self, result: Result<()>) {
+        if let Some(sender) = self.sender {
+            let _ = sender.send(result);
+        }
+    }
+}
+
+/// A handle to get the flush result.
+#[derive(Debug)]
+pub struct FlushHandle {
+    region_id: RegionId,
+    receiver: Receiver<Result<()>>,
+}
+
+impl FlushHandle {
+    /// Waits until the flush job is finished.
+    pub async fn wait(self) -> Result<()> {
+        self.receiver.await.context(WaitFlushSnafu {
+            region_id: self.region_id,
+        })?
+    }
+}
+
+/// Flush scheduler.
+pub struct FlushScheduler<S: LogStore> {
+    scheduler: LocalScheduler<FlushRequest<S>>,
+}
+
+pub type FlushSchedulerRef<S> = Arc<FlushScheduler<S>>;
+
+impl<S: LogStore> FlushScheduler<S> {
+    /// Returns a new [FlushScheduler].
+    pub fn new(config: SchedulerConfig, compaction_scheduler: CompactionSchedulerRef<S>) -> Self {
+        let handler = FlushHandler {
+            compaction_scheduler,
+        };
+        Self {
+            scheduler: LocalScheduler::new(config, handler),
+        }
+    }
+
+    /// Schedules a flush request and return the handle to the flush task.
+    ///
+    /// # Panics
+    /// Panics if `sender` of the `req` is not `None`.
+    pub fn schedule_flush(&self, mut req: FlushRequest<S>) -> Result<FlushHandle> {
+        assert!(req.sender.is_none());
+
+        let region_id = req.region_id();
+        let sequence = req.flush_sequence;
+        let (sender, receiver) = oneshot::channel();
+        req.sender = Some(sender);
+
+        let scheduled = self.scheduler.schedule(req)?;
+        // Normally we should not have duplicate flush request.
+        ensure!(
+            scheduled,
+            DuplicateFlushSnafu {
+                region_id,
+                sequence,
+            }
+        );
+
+        Ok(FlushHandle {
+            region_id,
+            receiver,
+        })
+    }
+}
+
+struct FlushHandler<S: LogStore> {
+    compaction_scheduler: CompactionSchedulerRef<S>,
+}
+
+#[async_trait::async_trait]
+impl<S: LogStore> Handler for FlushHandler<S> {
+    type Request = FlushRequest<S>;
+
+    async fn handle_request(
+        &self,
+        req: FlushRequest<S>,
+        token: BoxedRateLimitToken,
+        finish_notifier: Arc<Notify>,
+    ) -> Result<()> {
+        let compaction_scheduler = self.compaction_scheduler.clone();
+        common_runtime::spawn_bg(async move {
+            execute_flush(req, compaction_scheduler).await;
+
+            // releases rate limit token
+            token.try_release();
+            // notify scheduler to schedule next task when current task finishes.
+            finish_notifier.notify_one();
+        });
+
+        Ok(())
+    }
+}
+
+async fn execute_flush<S: LogStore>(
+    req: FlushRequest<S>,
+    compaction_scheduler: CompactionSchedulerRef<S>,
+) {
+    let mut flush_job = FlushJob {
```

Comment:
I'd like to impl  `From<FlushRequest<S>>` for `FlushJob`.

Reply:
`FlushRequest` and `FlushJob` are not equivalent, there is data loss during the conversion process. Is this suitable for `From`? Is `to_flush_job` will be better?

Reply:
`From<FlushRequest<S>>` consumes the request. I use `From<&FlushRequest<S>>` instead.